File: frontend/lib/application/rank_symbols/rank_symbols.dart
import '../../domain/symbol.dart';
import '../../domain/candle_series.dart';
import '../../domain/scoring/symbol_score_calculator.dart';
import 'ranked_symbol.dart';

/// Weight configuration for a score calculator.
class ScoreWeight {
  final SymbolScoreCalculator calculator;
  final double weight;
  const ScoreWeight({required this.calculator, required this.weight});
}

/// Interface for symbol ranking use case.
abstract class RankSymbols {
  List<RankedSymbol> rank(Map<AppSymbol, CandleSeries> seriesBySymbol);
}

/// Default implementation: deterministic, weighted, stable.
class RankSymbolsImpl implements RankSymbols {
  final List<ScoreWeight> scoreWeights;
  const RankSymbolsImpl({required this.scoreWeights});

  @override
  List<RankedSymbol> rank(Map<AppSymbol, CandleSeries> seriesBySymbol) {
    final List<RankedSymbol> ranked = [];
    for (final entry in seriesBySymbol.entries) {
      final symbol = entry.key;
      final series = entry.value;
      final Map<String, double> scores = {};
      double total = 0.0;
      for (final sw in scoreWeights) {
        if (sw.weight == 0) continue;
        final score = sw.calculator.scoreSeries(series);
        scores[sw.calculator.name] = score;
        total += score * sw.weight;
      }
      ranked
          .add(RankedSymbol(symbol: symbol, scores: scores, totalScore: total));
    }
    ranked.sort((a, b) {
      final cmp = b.totalScore.compareTo(a.totalScore);
      if (cmp != 0) return cmp;
      return a.symbol.value.compareTo(b.symbol.value);
    });
    return ranked;
  }
}

File: backend/application/usecases/rank_symbols.go
package usecases

import (
	"fmt"
	"pano_chart/backend/domain"
	"pano_chart/backend/domain/scoring"
	"sort"
)

// RankSymbols ranks symbols based on configured scoring strategies.
type RankSymbols interface {
	Rank(series map[domain.Symbol]domain.CandleSeries) ([]RankedSymbol, error)
}

type RankedSymbol struct {
	Symbol     domain.Symbol
	Scores     map[string]float64
	TotalScore float64
}

type ScoreWeight struct {
	Calculator scoring.SymbolScoreCalculator
	Weight     float64
}

// DefaultRankSymbols implements RankSymbols with explicit weights.
type DefaultRankSymbols struct {
	Weights []ScoreWeight
}

func NewDefaultRankSymbols(weights []ScoreWeight) *DefaultRankSymbols {
	return &DefaultRankSymbols{Weights: weights}
}

func (r *DefaultRankSymbols) Rank(series map[domain.Symbol]domain.CandleSeries) ([]RankedSymbol, error) {
	if len(series) == 0 {
		return nil, nil
	}

	result := make([]RankedSymbol, 0, len(series))
	for symbol, candles := range series {
		scores := make(map[string]float64)
		total := 0.0
		for _, w := range r.Weights {
			if w.Weight == 0 {
				continue
			}
			score, err := w.Calculator.Score(candles)
			if err != nil {
				return nil, fmt.Errorf("calculator %s failed for %s: %w", w.Calculator.Name(), symbol.String(), err)
			}
			scores[w.Calculator.Name()] = score
			total += score * w.Weight
		}
		result = append(result, RankedSymbol{
			Symbol:     symbol,
			Scores:     scores,
			TotalScore: total,
		})
		fmt.Printf("[RankSymbols] Calculated scores for %s: %+v, total: %.4f\n", symbol.String(), scores, total)
	}

	sort.Slice(result, func(i, j int) bool {
		if result[i].TotalScore == result[j].TotalScore {
			return result[i].Symbol.String() < result[j].Symbol.String()
		}
		return result[i].TotalScore > result[j].TotalScore
	})
	return result, nil
}
